### ISSUE TITLE
Increasing benchmark duration results in degraded performance / failures

### ISSUE THREAD

**Issue by johnclarke96** [MystenLabs/narwhal#995]
### Description
Hi, I've been tinkering with narwhal and running benchmarks using the fabric tool to analyze performance over longer periods of time.

### Steps to reproduce
Running `fab local` for an increasing duration results in degraded TPS, BPS, and latency. At long enough durations, the round fails to progress. In the failure case, it seems certain tokio channels are getting fully saturated and becoming a bottleneck, though I have just begun an investigation.

See below table for the experiments I ran on an m5.metal.

https://gist.github.com/johnclarke96/02c30ec769e8307f877e32d38024dcfa

Is it a known issue that performance degrades when the benchmark is run for extended periods of time? Is narwhal expected to fail at long enough durations?

**Comment by mwtian** [comment]
Hi John, thanks for the detailed issue description. We are investigating problems related to clogged tokio channels at the moment. Tagging @huitseeker @akichidis @asonnino.

**Comment by huitseeker** [comment]
I have a partial reproduction of this, with a significantly different cutoff point given that I used a different machine, with different hardware characteristics.

We're recently added an executor at the end of consensus, which retrieves data from all workers, and streams it to a downstream client. Even though in benchmarks, the subscriber is a no-op subscriber, the data is still fetched and streamed, and that should arguably not be part of benchmarks.

I've taken the following commit for measures:
```
huitseeker@sleipnir➜tmp/narwhal/benchmark(main✗)» git rev-parse HEAD
a08957f5b1f06c174e32c367f60ab502428f53c4
huitseeker@sleipnir➜tmp/narwhal/benchmark(main✗)» git log -n1
commit a08957f5b1f06c174e32c367f60ab502428f53c4 (HEAD -> main, origin/main, origin/HEAD)
Author: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Date:   Tue Sep 20 01:48:39 2022 +0000

    chore(deps): bump hdrhistogram from 7.5.1 to 7.5.2 (#1013)
```
Here are the results (on a Linux x86_64 w/ Ryzen 9 3950X, 64GB):
https://gist.github.com/huitseeker/8f65b123a000057e628e43ceb0a13e79

**Comment by asonnino** [comment]
> We're recently added an executor at the end of consensus, which retrieves data from all workers, and streams it to a downstream client. Even though in benchmarks, the subscriber is a no-op subscriber, the data is still fetched and streamed, and that should arguably not be part of benchmarks: the primary retrieving this data can become a bottleneck.

I agree, this executor should not be part of the benchmark as it will likely be the bottleneck (for no real reason). If we keep it, the benchmarks will be less useful to us since they will not be able to detect when new changes decrease performance

**Comment by maks-ivanov** [comment]
I certainly agree that perf metrics should be disaggregated between consensus and execution to allow for a more granular view, but I'd like to push back on the notion that executor is not a "real" bottleneck that should be excluded entirely. If the benchmark metric shows result X but there is no way for transactions to get retrieved fast enough from the mempool to match this rate, then I don't see how one could treat it as a real performance indicator in practice. In a sense it doesn't matter how fast the executor is, the performance limitation is still coming from narwhal architecture. I'm not a distributed systems expert though so open to suggestions / other ways of viewing this issue

**Comment by asonnino** [comment]
I think we are not talking about the same type of benchmark. If we want a benchmark to know the "real" performance of the system then we should have the executor crate and the execution engine. I understand that what we want here is different: we want a way to ensure our future changes do not make our performance regress. So choking on a known bottleneck will defeat that purpose (finding an efficient way to retrieve Narwhal's block is on our TODO list).

**Comment by maks-ivanov** [comment]
That's reasonable, I have some more data to add - I produced a scenario where the executor can fetch and process a block at roughly the same rate as consensus outputs them, and the network still stopped progressing after some time. So it looks like a problem but not necessarily *the* problem

**Comment by asonnino** [comment]
Thanks for the insight! We suspect we have an issue with our synchroniser, and your experiments seem to point into that direction as well
